File: typescript/gql-codegen/src/schema/client/directive.ts
import { z } from 'zod/v4';
import { inputFieldSchema } from '../shared.js';

export const directiveSchema = z.object({
    name: z.string(),
    arguments: z.record(z.string(), inputFieldSchema),
    locations: z.array(z.enum([
        'QUERY',
        'MUTATION',
        'SUBSCRIPTION',
        'FIELD',
        'FRAGMENT_DEFINITION',
        'FRAGMENT_SPREAD',
        'INLINE_FRAGMENT',
        'VARIABLE_DEFINITION',
    ]))
})

File: rust/gqlcodegen/src/schema/client/directive.rs
use std::collections::HashMap;

use crate::schema::shared::InputField;

#[derive(Debug, serde::Deserialize)]
pub enum DirectiveLocation {
    #[serde(rename(deserialize = "QUERY"))]
    Query,
    #[serde(rename(deserialize = "MUTATION"))]
    Mutation,
    #[serde(rename(deserialize = "SUBSCRIPTION"))]
    Subscription,
    #[serde(rename(deserialize = "FIELD"))]
    Field,
    #[serde(rename(deserialize = "FRAGMENT_DEFINITION"))]
    FragmentDefinition,
    #[serde(rename(deserialize = "FRAGMENT_SPREAD"))]
    FragmentSpread,
    #[serde(rename(deserialize = "INLINE_FRAGMENT"))]
    InlineFragment,
    #[serde(rename(deserialize = "VARIABLE_DEFINITION"))]
    VariableDefinition,
}

#[derive(Debug, serde::Deserialize)]
pub struct Directive {
    name: String,
    arguments: HashMap<String, InputField>,
    locations: Vec<DirectiveLocation>
}
